_posts/2021-12-09-cdk-cr.md

```diff
@@ -0,0 +1,174 @@
+
+The benefit is that you don't need to build a lambda yourself, which reduces the risk of missing the `cfn_response` callback and reduces the verbosity of resources that you need to define. It's syntactic sugar, but it also helps to create safe and dependable code.
+
+## How do I use the AwsCustomResource construct?
+
+Now that we understand the benefits, let's learn how to use the module.
+
+Start by importing `custom_resources`:
+
+```py
+from aws_cdk import (
+    custom_resources as cr
+)
+```
+
+The minimal set of properties for creating a custom resource is as follows:
+
+```py
+cr.AwsCustomResource(self, 'MyResource',
+                     policy=custom-resource_policy,
+                     on_create=on_create_call,
+                     on_update=on_update_call,
+                     on_delete=on_delete_call)
+```
+
+Under the hood this resource will create a lambda, execution role, policies and custom resource provider for you. But that's all hidden away from you. You don't need to write any lambda code yourself.
+
+The `policy` parameter allows you to define the permissions assigned to the underlying lambda function. The module provides a couple of convenient ways to provide this.
+
+Firstly, you can tell it to infer the permissions from the SDK calls made:
+
+```py
+policy=cr.AwsCustomResourcePolicy.from_sdk_calls(
+    resources=cr.AwsCustomResourcePolicy.ANY_RESOURCE
+)
+```
+
+Otherwise you can provide a full policy document yourself. here's an example for a IoT role alias resource:
+
+```py
+policy=cr.AwsCustomResourcePolicy.from_statements(
+    statements=[iam.PolicyStatement(
+        effect=iam.Effect.ALLOW,
+        actions=[
+            "iot:CreateRoleAlias",
+            "iot:UpdateRoleAlias",
+            "iot:DeleteRoleAlias"
+        ],
+        resources=["*"]
+    ), iam.PolicyStatement(
+        effect=iam.Effect.ALLOW,
+        actions=[
+            "iam:PassRole"
+        ],
+        resources=[role_arn]
+    )]
+)
+```
+
+For each of the resource events — `on_create`, `on_update` and `on_delete` you can provide a `AwsSdkCall` object which defines which AWS SDK gets called.
+
+For example, to create an IoT Thing Group on the create event:
+
+```py
+on_create=cr.AwsSdkCall(
+    action="createThingGroup",
+    service="Iot",
+    parameters={
+        'thingGroupName': thing_group_name
+    },
+    physical_resource_id=cr.PhysicalResourceId.of(thing_group_name)
+)
+```
+
+This will call the [`createThingGroup` AWS SDK method](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Iot.html#createThingGroup-property) at resource creation time and pass parameters.
+
+`physical_resource_id` is a way of uniquely identifying the resource created at runtime. In most cases you would use the ARN of the object, which you can get from the response, but you may use something else.
```

Comment:
Link to more references here. What does CFN actually use the ID for?

Reply:
It's the most important thing in the custom resource because it determines how update/delete behaviour works.

If a physical resource ID changes it - its meant to signal that a new resource has been created. So if you were to update the cfn resource, and send back a new physical resource ID - you've told cloudformation that you made a new one, and the old one needs to be deleted. If the stack fails and needs to roll back - it will delete the new resource. If it succeeds, it will delete the old resource. And it will send new "signals" to the custom resource handler to do so. e.g. update your cloudformation template and you will get a) a signal to 'update' the resource and b) a signal to delete some other resource. If you don't send a new physical ID - the implication is you did an in-place update.

This is critical to understand - because some things can't be updated in place (e.g. domain names in an ACM certificate - that would require generating an entirely new certificate). Other things have similar limitations because their underlying ID's and names are referenced elsewhere (IAM roles/users etc).

Reply:
So if I understand this right -- this allows the "logical resource" to act as a facade to the physical happenings? The CR maintainer can control the physical via the resource ID and hide that detail away?

To continue your ACM example -- if I update the property in my CFN template (logical update), the physical delete and re-creation (replacement) is opaque to me as the user of the custom resource?

Reply:
yes - the logical resource is effectively the definition of the resource in your template - the logical ID is whatever name you give it.

The physical resource ID can be whatever you want it to be.

And yes - it will be opaque to the user. But it's important from a functionality standpoint because how you do that implementation is going to determine whether or not an update has consequences (e.g. downtime)

Reply:
Thanks for clarifying. I wish the AWS docs made it as clear as you did 😆 .

Reply:
Added text to explain this